File: educational-core/src/com/jetbrains/edu/learning/courseFormat/ItemContainer.java
package com.jetbrains.edu.learning.courseFormat;

import com.jetbrains.edu.learning.EduUtils;
import one.util.streamex.StreamEx;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class ItemContainer extends StudyItem {
  protected List<StudyItem> items = new ArrayList<>();

  @Nullable
  public StudyItem getItem(@NotNull final String name) {
    return StreamEx.of(items).findFirst(item -> item.getName().equals(name)).orElse(null);
  }

  @NotNull
  public List<StudyItem> getItems() {
    return Collections.unmodifiableList(items);
  }

  public void setItems(List<StudyItem> items) {
    this.items = items;
  }

  public void addItem(StudyItem item) {
    items.add(item);
  }

  public void removeItem(StudyItem item) {
    items.remove(item);
  }

  public void sortItems() {
    Collections.sort(items, EduUtils.INDEX_COMPARATOR);
  }
}

File: edu-format/src/com/jetbrains/edu/learning/courseFormat/ItemContainer.kt
package com.jetbrains.edu.learning.courseFormat

abstract class ItemContainer : StudyItem() {
  var items: List<StudyItem>
    get() = _items
    set(value) {
      _items = value.toMutableList()
    }

  private var _items = mutableListOf<StudyItem>()
  override fun init(parentItem: ItemContainer, isRestarted: Boolean) {
    parent = parentItem
    for ((i, item) in items.withIndex()) {
      item.index = i + 1
      item.init(this, isRestarted)
    }
  }

  fun getItem(name: String): StudyItem? {
    return items.firstOrNull { it.name == name }
  }

  fun addItem(item: StudyItem) {
    _items.add(item)
  }

  fun addItem(index: Int, item: StudyItem) {
    _items.add(index, item)
  }

  fun replaceItem(existingItem: StudyItem, newItem: StudyItem) {
    val index = _items.indexOf(existingItem)
    if (index < 0) return
    _items[index] = newItem
  }

  fun removeItem(item: StudyItem) {
    _items.remove(item)
  }

  open fun sortItems() {
    _items.sortBy { it.index }
  }
}
